tuf/repository_lib.py

```diff
@@ -1541,6 +1559,168 @@ def _get_hashes_and_length_if_needed(use_length, use_hashes, full_file_path,
 
 
 
+
+# I couldn't find a currently maintained python library for this, so
+# implementing it here. It would be better to implement this in c,
+# even better to use an existing library
+# This is inspired by: https://www.literateprograms.org/miller-rabin_primality_test__python_.html
+def miller_rabin_round(a, s, d, n):
+  a_to_power = pow(a, d, n)
+  if a_to_power == 1:
+    return True
+  for i in range(s):
+    if a_to_power == n - 1:
+      return True
+    a_to_power = (a_to_power * a_to_power) % n
+  return False
+
+
+
+
+
+def miller_rabin(n, rounds):
+  if n == 1:
+    return false
+  if n == 2 or n == 3:
+    return true
+
+  d = n -1
+  s = 0
+  while d % 2 == 0:
+    d = d >> 1
+    s = s + 1
+
+  for i in range(rounds):
+    a = random.randrange(n)
+    if not miller_rabin_round(a, s, d, n):
+      return False
+  return True
+
+
+
+
+# RSA Accumulator code insprired by https://github.com/ElrondNetwork/elrond-go/blob/v1.0.30/crypto/accumulator/rsa/rsaAcc.go
+def hash_to_prime(data):
+  # TODO: move constant definitions
+  basesMillerRabin = 12
+
+  h = blake2b(str(data).encode('utf-8'))
```

Comment:
Big issue: this only gives you a 512 bit number, so we're using 1 out of 2^{2048 -512} = 2^{1536}th of the possible space!

Need to use an XOF to get >512 bits. Also, we want uniformly random values from [0, MODULUS).

See, e.g., [[Lem18]](https://arxiv.org/abs/1805.10941)

I'd recommend to implement as (approximate):

```python
def hash_to_prime(data, modulus):
  modulus_bits = math.ceil(math.log2(modulus))  # TODO: may need to round this up to the nearest byte

  # a number n has about 1/ln(n) chance of being prime so we may actually want to make this depend on our modulus
  for _ in range(MAX_ATTEMPTS):
    digest: bytes = xof(data).digest(modulus_bits)  # might be > modulus, so we need to wrap
    value = int.from_bytes(digest, 'big') # endianness arbitrary but must be the same everywhere

    # value should be in [0, modulus)
    if value >= modulus:
      # rejection sampling: this happens up to 1/2 of the time if modulus is 
      # Lemire gives a better method (alg. 5): https://arxiv.org/pdf/1805.10941.pdf
      continue

    if not miller_rabin(value, MILLER_RABIN_ROUNDS):
      continue  # not prime
  return value
```